### PR TITLE
[WIP] receiver: Split

### PR DESCRIPTION
**Pull request by Chans321**
Signed-off-by: Chans321 <tsschand@gmail.com>

* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

### CODE REVIEW

File: pkg/route/handler.go [hunk 1]
```diff
@@ -0,0 +1,668 @@
...
+
+// writeQuorum returns minimum number of replicas that has to confirm write success before claiming replication success.
+func (h *Handler) writeQuorum() int {
+	return int((h.options.ReplicationFactor / 2) + 1)
+}
+
+// fanoutForward fans out concurrently given set of write requests. It returns status immediately when quorum of
+// requests succeeds or fails or if context is canceled.
+func (h *Handler) fanoutForward(pctx context.Context, tenant string, replicas map[string]replica, wreqs map[string]*prompb.WriteRequest, successThreshold int) error {
+	var errs terrors.MultiError
```
Comment (brancz) [inline pkg/route/handler.go:367]: This needs to be changed to `errutil.MultiError`

File: pkg/route/handler.go [hunk 2]
```diff
@@ -0,0 +1,668 @@
...
+	return nil
+}
+
+
+// countCause counts the number of errors within the given error
+// whose causes satisfy the given function.
+// countCause will inspect the error's cause or, if the error is a MultiError,
+// the cause of each contained error but will not traverse any deeper.
+func countCause(err error, f func(error) bool) int {
+	errs, ok := err.(terrors.MultiError)
```
Comment (brancz) [inline pkg/route/handler.go:580]: This needs to be changed to `errutil.MultiError`

**Review by brancz — COMMENTED** [review]
I think this is pretty close! Nice work!

**Comment by yashrsharma44** [comment]
What are the things that are blocking this PR? Just curious. 🤔

**Comment by bwplotka** [comment]
Rebase, would be one thing 🤗

**Comment by brancz** [comment]
It just needs to be squashed + rebased I think. If @Chans321 doesn't mind I guess you could take it over.

**Comment by yashrsharma44** [comment]
Sure, I can get this PR completed.

**Comment by bwplotka** [comment]
As promised, the modification of the proposal is ready: https://github.com/thanos-io/thanos/pull/4036 🤗  PTAL before moving with this PR.

**Comment by kakkoyun** [comment]
Thanks a million for amazing work @Chans321 🎉 

Deprecated by https://github.com/thanos-io/thanos/pull/4231

**Pull request closed without merging**